### ISSUE TITLE
Postgres recovery handlers can overflow stack

### ISSUE THREAD

**Issue by gstro** [tpolecat/doobie#512]
Using Postgres recovery combinators recursively can overflow the stack.  

For example, running the following:
```scala
def retryInsert(insertId: ConnectionIO[Long], retries: Int): ConnectionIO[Long] = 
  insertId.onUniqueViolation(retryInsert(insertId, retries - 1))
```
on a column with a `UNIQUE` constraint has the following output:
```scala
scala> retryInsert(testInsert.query[Long].unique, 10).transact(xa).unsafePerformSync
java.lang.StackOverflowError
  at scalaz.FreeInstances.freeMonad(Free.scala:424)
  at .retryInsert(<console>:46)
  at .retryInsert(<console>:46)
  at .retryInsert(<console>:46)
  ...
```

**Comment by edmundnoble** [comment]
@gstro You forgot to compare retries to zero and terminate.

**Comment by tpolecat** [comment]
@edmundnoble yeah, but infinite retry should still work (and retry 1000 times shouldn't construct the whole chain eagerly).

**Comment by edmundnoble** [comment]
Doesn't seem to construct the chain eagerly because ConnectionIO is Free. However, it sounds like you need a `FreeMonadError`, like the `IO` type in cats-effect (see the `AndThen` constructor).

**Comment by gstro** [comment]
@edmundnoble You're right about the termination, though I got the same result when that was included:
```scala
def retryInsert(insertId: ConnectionIO[Long], retries: Int): ConnectionIO[Long] =
  if (retries == 0) insertId
  else insertId.onUniqueViolation(retryInsert(insertId, retries - 1))
```

The fix in #513 worked for me.
